drivers/pwm/pwm_ambiq_ctimer.c

```diff
@@ -0,0 +1,248 @@
+	uint32_t timer_seg;
+	uint32_t pwm_type;
+	uint32_t clock_sel;
+	const struct pinctrl_dev_config *pincfg;
+};
+
+static uint32_t get_clock_cycles(uint32_t clock_sel)
+{
+	uint32_t ret = 0;
+
+	switch (clock_sel) {
+	case 1:
+		ret = 12000000;
+		break;
+	case 2:
+		ret = 3000000;
+		break;
+	case 3:
+		ret = 187500;
+		break;
+	case 4:
+		ret = 47000;
+		break;
+	case 5:
+		ret = 12000;
+		break;
+	case 6:
+		ret = 32768;
+		break;
+	case 7:
+		ret = 16384;
+		break;
+	case 8:
+		ret = 2048;
+		break;
+	case 9:
+		ret = 256;
+		break;
+	case 10:
+		ret = 512;
+		break;
+	case 11:
+		ret = 32;
+		break;
+	case 12:
+		ret = 1000;
+		break;
+	case 13:
+		ret = 116;
+		break;
+	case 14:
+		ret = 100;
+		break;
+	case 15: /* todo: check on value */
+		ret = 0;
+		break;
+	case 16:
+		ret = 8192;
+		break;
+	case 17:
+		ret = 4096;
+		break;
+	case 18:
+		ret = 1024;
+		break;
+	default:
+		ret = 12000000;
+		break;
+	}
+
+	return ret;
+}
+
+static void start_clock(uint32_t clock_sel)
+{
+	switch (clock_sel) {
+	default:
+		break;
+	case 1:
+	case 2:
+	case 3:
+	case 4:
+	case 5:
+		am_hal_clkgen_control(AM_HAL_CLKGEN_CONTROL_SYSCLK_MAX, 0);
+		break;
+	case 6:
+	case 7:
+	case 8:
+	case 9:
+	case 14: /* RTC will assume XTAL since LFRC isn't as accurate */
+	case 16:
+	case 17:
+	case 18:
+		am_hal_clkgen_control(AM_HAL_CLKGEN_CONTROL_XTAL_START, 0);
+		break;
+	case 10:
+	case 11:
+	case 12:
+	case 13:
+		am_hal_clkgen_control(AM_HAL_CLKGEN_CONTROL_LFRC_START, 0);
+		break;
+	}
+}
+
+static int ambiq_timer_pwm_set_cycles(const struct device *dev, uint32_t channel,
+				      uint32_t period_cycles, uint32_t pulse_cycles,
+				      pwm_flags_t flags)
+{
+	const struct pwm_ambiq_timer_config *config = dev->config;
+
+	if (period_cycles == 0) {
+		LOG_ERR("period_cycles can not be set to zero");
+		return -ENOTSUP;
+	}
+
+	if (flags & PWM_POLARITY_INVERTED) {
+		if (pulse_cycles == 0) {
+			/* make pulse cycles greater than period so event never occurs */
+			pulse_cycles = period_cycles + 1;
+		} else {
+			pulse_cycles = period_cycles - pulse_cycles;
+		}
+	} else {
+		if (pulse_cycles == period_cycles) {
+			--pulse_cycles;
+		} else if (pulse_cycles == 0) {
+			period_cycles = 0;
+			pulse_cycles = 1;
+		}
+	}
+
+	uint32_t seg = 0;
+
+	if (config->timer_seg == 0) {
+		seg = 0x0000FFFF;
+	} else if (config->timer_seg == 1) {
+		seg = 0xFFFF0000U;
+	} else {
+		seg = 0xFFFFFFFFU;
+	}
+
+	/* todo: need to check if all of this is required */
+	am_hal_ctimer_clear(config->timer_num, seg);
+	am_hal_ctimer_period_set(config->timer_num, seg, period_cycles, pulse_cycles);
+	am_hal_ctimer_start(config->timer_num, seg);
+
+	return 0;
+}
+
+static int ambiq_timer_pwm_get_cycles_per_sec(const struct device *dev, uint32_t channel,
+					      uint64_t *cycles)
+{
+	struct pwm_ambiq_timer_data *data = dev->data;
+	int err = 0;
+
+	/* clean up upper word of return parameter */
+	*cycles &= 0xFFFFFFFFU;
+
+	/* cycles of the timer clock */
+	*cycles = data->cycles;
```

Comment:
What's the point in clearing the upper word when in the next step the whole value is overwritten?